rust/src/sdp/logger.rs

```diff
@@ -0,0 +1,153 @@
+    }
+    if let Some(email) = &msg.email {
+        js.set_string("email", email)?;
+    }
+    if let Some(phone_number) = &msg.phone_number {
+        js.set_string("phone_number", phone_number)?;
+    }
+    if let Some(conn_data) = &msg.connection_data {
+        log_connection_data(conn_data, js)?;
+    }
+    if let Some(bws) = &msg.bandwidths {
+        log_bandwidth(bws, js)?;
+    }
+    js.set_string("time", &msg.time)?;
+    if let Some(repeat_time) = &msg.time_zone {
+        js.set_string("repeat_time", repeat_time)?;
+    }
+    if let Some(tz) = &msg.time_zone {
+        js.set_string("timezone", tz)?;
+    }
+    if let Some(enc_key) = &msg.encryption_key {
+        js.set_string("encryption_key", enc_key)?;
+    }
+    if let Some(attrs) = &msg.attributes {
+        log_attributes(attrs, js)?;
+    }
+    if let Some(media) = &msg.media_description {
+        log_media_description(media, js)?;
+    }
+    Ok(())
+}
+
+fn log_media_description(
+    media: &Vec<MediaDescription>, js: &mut JsonBuilder,
+) -> Result<(), JsonError> {
+    if !media.is_empty() {
+        js.open_array("media_descriptions")?;
+        for m in media {
+            js.start_object()?;
+            let port = if let Some(num_ports) = m.number_of_ports {
+                format!("{}/{}", m.port, num_ports)
+            } else {
+                format!("{}", m.port)
+            };
+            let mut media = format!("{} {} {}", &m.media, &port, &m.proto);
+            for f in &m.fmt {
+                media = format!("{} {}", media, f);
+            }
+            js.set_string("media", &media)?;
+
+            if let Some(session_info) = &m.session_info {
+                js.set_string("media_info", session_info)?;
+            };
+            if let Some(bws) = &m.bandwidths {
+                log_bandwidth(bws, js)?;
+            }
+            if let Some(conn_data) = &m.connection_data {
+                log_connection_data(conn_data, js)?;
+            }
+            if let Some(attrs) = &m.attributes {
+                log_attributes(attrs, js)?;
+            }
+            js.close()?;
+        }
+    }
+    js.close()?;
+
+    Ok(())
+}
+
+fn log_bandwidth(bws: &Vec<String>, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    if !bws.is_empty() {
+        js.open_array("bandwidths")?;
+        for bw in bws {
+            js.append_string(bw)?;
+        }
+        js.close()?;
+    }
+    Ok(())
+}
+
+fn log_connection_data(conn_data: &ConnectionData, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    let mut conn = format!(
+        "{} {} {}",
+        &conn_data.nettype,
+        &conn_data.addrtype,
+        &conn_data.connection_address.to_string()
+    );
+    if let Some(ttl) = conn_data.ttl {
+        conn = format!("{}/{}", conn, ttl);
+        js.set_uint("ttl", ttl as u64)?;
+    }
+    if let Some(num_addrs) = conn_data.number_of_addresses {
+        conn = format!("{}/{}", conn, num_addrs);
```

Comment:
How do I distinguish the cases ttl and no number_of addresses, vs the case no ttl and number_of_addresses ?

Reply:
The TTL field is present only for IPv4 multicast addresses.
However, its  may still not be very clear.